python/GafferDispatch/FrameMask.py

```diff
@@ -0,0 +1,65 @@
+##########################################################################
+#
+#  Copyright (c) 2018, Image Engine Design Inc. All rights reserved.
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions are
+#  met:
+#
+#      * Redistributions of source code must retain the above
+#        copyright notice, this list of conditions and the following
+#        disclaimer.
+#
+#      * Redistributions in binary form must reproduce the above
+#        copyright notice, this list of conditions and the following
+#        disclaimer in the documentation and/or other materials provided with
+#        the distribution.
+#
+#      * Neither the name of John Haddon nor the names of
+#        any other contributors to this software may be used to endorse or
+#        promote products derived from this software without specific prior
+#        written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+##########################################################################
+
+import IECore
+
+import Gaffer
+import GafferDispatch
+
+class FrameMask( GafferDispatch.TaskNode ) :
```

Comment:
I find it really hard to imagine SceneFrameMask and ImageFrameMask for some reason, and I'm not aware of anyone every having asked for it. Perhaps this is because in many cases the "enabled" plug can be used for a similar purpose?

I suggest we stick with the simpler FrameMask for now, knowing that we can easily use a compatibility config if we need to in the future.

Reply:
> I find it really hard to imagine SceneFrameMask and ImageFrameMask for some reason, and I'm not aware of anyone every having asked for it.

See the IE published Box named "frameHold" which is essentially an ImageFrameMask. I coudn't find a published version for scenes, but I've definitely sent people the "SceneContextVariables with a variable called 'frame' and an expression driving the value" example many times.

Reply:
Network issues are preventing me checking those out, but holding a frame and masking it are different things aren't they? Holding means the image/scene still exists, but is taking it's data from a different upstream frame. Masking means the task no longer exists at all. I'm going to stick to my guns on this one, with the fallback that I'm perfectly happy to rename the node if the need arises in the future...

> I've definitely sent people the "SceneContextVariables with a variable called 'frame' and an expression driving the value" example many times.

Any reason you're not sending them a SceneTimeWarp?

Reply:
> holding a frame and masking it are different things aren't they? Holding means the image/scene still exists, but is taking it's data from a different upstream frame.

Yes, but inverting your expression turns the mask into a hold, right? I guess its better described as a frame clamp? The use case here is to dispatch an enormous range (0-3000) and use those boxes to make sure certain branches don't render outside a start/stop frame range. So they're defining the frames to keep rather than the frames to skip.

> Any reason you're not sending them a SceneTimeWarp?

I guess because its easier to tell someone "make an expression that drives the frame however you want" than to do the same for speed and offset.

Reply:
> Yes, but inverting your expression turns the mask into a hold, right? I guess its better described as a frame clamp?

Sorry, I'm not following, maybe just because I've been unable to see the frameHold example. To me, "hold" means a particular upstream frame is substituted in for some downstream frames. For instance, all frames greater than 10 might read frame 10 instead, so frame 10 is repeated. And "mask" means to ignore the tasks for some frames entirely. If you invert the mask, you're just ignoring a different set of frames, not holding anything. Nothing is repeated.

> The use case here is to dispatch an enormous range (0-3000) and use those boxes to make sure certain branches don't render outside a start/stop frame range. So they're defining the frames to keep rather than the frames to skip.

Yeah, that's why the FrameMask node accepts the list of frames to keep.

Reply:
I think I'm mixing things up... lets just end this discussion, its gotten maddening.